# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Hazammm/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

I'm choosing Lane 2: Refresh/Content Opportunity Scoring. This lane directly builds on the Week 1 starter notebooks, where I already saw a hand-written rule (Precision@50 ≈ 0.24) get beaten by a learned model (≈0.74). I want to go deeper into which pages deserve review first and why, using reason codes a human can actually inspect.

In [4]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

Question: Given a page's search and engagement signals, how do we rank which pages a content reviewer should look at first for a refresh?

Unit of analysis: one page (content item) per client.

Decision: which pages, out of potentially thousands, does the reviewer spend their limited time on first?

Action: the reviewer refreshes, expands, or otherwise edits the flagged page — or leaves it if it's a false positive.

Cost of a wrong call:

False positive (flagged but page was fine) → wasted reviewer time
False negative (a genuinely declining page not flagged) → traffic keeps dropping unnoticed

Why ML helps: hand-written rules treat every signal equally and can't learn combinations (e.g. "old AND declining AND still gets traffic" matters more than any single factor). A model can weigh signals the way the data actually shows they matter, and rank pages far better than a fixed rule — as shown by the ~3x precision lift already observed in the starter pipeline.

In [5]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [6]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


In [7]:
# Clone your repo so the notebook can see the data file
!git clone https://github.com/Hazammm/flyrank-ml-internship.git repo
%cd repo

Cloning into 'repo'...
remote: Enumerating objects: 122, done.
remote: Counting objects: 100% (122/122), done.
remote: Compressing objects: 100% (78/78), done.
remote: Total 122 (delta 34), reused 97 (delta 28), pack-reused 0 (from 0)
Receiving objects: 100% (122/122), 1.89 MiB | 13.45 MiB/s, done.
Resolving deltas: 100% (34/34), done.
/content/repo/repo


In [8]:
import pandas as pd

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

# Filter same way the starter pipeline does: impressions_90d > 0, content_age_days >= 90
df = df[(df["impressions_90d"] > 0) & (df["content_age_days"] >= 90)]

print(f"Total pages after filtering: {len(df)}")

decline_pct = (df["trend_direction"] == "down").mean() * 100
print(f"Percent of pages declining: {decline_pct:.1f}%")

stale_visible = ((df["days_since_last_update"] >= 180) & (df["impressions_90d"] >= 500)).sum()
print(f"Stale-but-visible pages (haven't been updated in 6+ months, still getting traffic): {stale_visible}")

Total pages after filtering: 30000
Percent of pages declining: 54.2%
Stale-but-visible pages (haven't been updated in 6+ months, still getting traffic): 17


Out of 30,000 pages (after filtering for pages with real impressions and at least 90 days of age), 54.2% are currently declining — meaning a random or naive approach would flag right pages barely better than a coin flip. More importantly, 17 pages are both stale (not updated in 6+ months) and still getting meaningful traffic (500+ impressions) — these are concrete, low-risk opportunities: pages that are clearly worth reviewing but have been sitting untouched. This shows there's a real, actionable gap for a smarter ranking system to catch that a person scanning manually might miss.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

What I can claim:
This data shows observed patterns and correlations — for example, that a page's age, freshness, or search position relates to whether it's declining or how it performs. These are real, measurable relationships in this 30,000-page anonymized sample.

What I cannot claim:

I cannot claim this proves how Google's ranking algorithm works — I only see the outcome (position, clicks, impressions), not the algorithm itself.
I cannot claim that refreshing a flagged page will guarantee a recovery in traffic — that would require running an actual experiment (e.g. refresh some pages, leave others alone, compare results), not just analyzing existing data.
I cannot claim causation from correlation — e.g. even if declining pages tend to be older, that doesn't mean age causes the decline; something else (like the topic going out of relevance) could explain both.

What my output actually is:
A ranked, decision-support list — a way to help a human reviewer prioritize limited time, not a guarantee of what will happen to any single page. Every claim I make should be phrased as "observed," "associated with," or "directional" — never "proven" or "caused."

In [9]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## Self-check

Before you submit, confirm each line honestly:

- [x] Every section above is filled — markdown thinking AND the code that backs it
- [x] The notebook runs top to bottom with no errors (Runtime → Run all)
- [x] No client names, URLs, or private queries anywhere
- [x] My claims use careful words: observed, measured, directional, decision-support
- [x] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.